impeller/docs/babys_first_triangle.md

```diff
@@ -0,0 +1,276 @@
+
+Patching the vertex and fragment shaders to provide data for the second color per vertex and passing that along to the fragment shader as a varying is left as an exercise to the reader. We just discussed it in the section above.
+
+In the fragment shader (where we need to the current time in seconds), add a uniform block like so:
+
+```glsl
+uniform FragInfo {
+  float time;
+} frag_info;
+```
+
+The compiler will helpfully spit out a C++ struct that looks similar. Like the `PerVertexData`, the compiler knows the alignment of the various fields and will place them with the right padding as necessary. All you need to do is fill this in in C++ code and provide that data to GPU as a device buffer. Let's take a peek at what the compiler generated for us:
+
+```c++
+struct FragInfo {
+  Scalar time; // (offset 0, size 4)
+}; // struct FragInfo (size 4)
+```
+
+Seems pretty straightforward. Just to prove a point, I'll add a few more fields that I know need padding. [Tinker around in the compiler explorer](https://tinyurl.com/23q4o4cf) to learn more about what the compiler generates.
+
+```glsl
+uniform FragInfo {
+  float time;
+  vec2 bar;
+  vec4 baz;
+} frag_info;
+```
+
+in GLSL generates the following C++ struct. Notice the additional padding you didn't have to think about.
+
+```c++
+struct FragInfo {
+  Scalar time;               // (offset 0, size 4)
+  Padding<4> _PADDING_bar_;  // (offset 4, size 4)
+  Point bar;                 // (offset 8, size 8)
+  Vector4 baz;               // (offset 16, size 16)
+};                           // struct FragInfo (size 32)
+```
+
+In our playground, create this `FragInfo` struct and set the time value as the current value in seconds.
+
+```c++
+FS::FragInfo frag_info;
+frag_info.time = fml::TimePoint::Now().ToEpochDelta().ToSecondsF();
+```
+
+We need to place this in GPU accessible memory. Here, we will use another utility called the host buffer.
+
+#### Host & Device Buffers
+
+Unlike `malloc` buffers that can be resized via `realloc`, device buffers cannot be resized in Impeller. Also, it is bad form to create many small device buffers. For this reason, Impeller prefers to stage all data (uniform, vertex, or otherwise) into one large allocation. Then at draw time, each draw call references information at a specific offset and length into that larger allocation.
+
+An easy to way to achieve this scheme is to use a `HostBuffer`. A host buffer is a buffer allocated on the heap (using `malloc` or similar) whose main usage is to grow as quickly as possible. Impeller stages all data for the frame in such a buffer. As the buffer is being constructed, views (offset and length) into this buffer are noted in the command stream. Just before the draw call is submitted, this information is uploaded to the GPU at which time with buffers being reused if necessary.
```

Comment:
primarily this uses the driver's allocator, VMA, Metal allocator, or malloc on GLSL where there is no allocation API

Reply:
Yeah, I was hand wavy about the implementation details since they're so different and we keep tinkering on them anyway.